### PR TITLE
Added new topic for create custom rest api

### LINKED ISSUE #9117: New topic for creating custom REST Api
# New topic request

## Description

In the rest tutorial documentation, there is no example to create a custom rest API

We can add a new topic under the rest tutorials with examples

### PR DESCRIPTION
**Pull request by nitish-ranjan05**
## Purpose of this pull request

Fixes #9117

## Description

-  This page describes how you can create a custom REST API in Magento 2.

## Affected DevDocs pages

- https://devdocs.magento.com/guides/v2.3/rest/tutorials/create-custom-rest-api.html
- https://devdocs.magento.com/guides/v2.4/rest/tutorials/create-custom-rest-api.html

whatsnew
Added the [Create a custom REST API](https://devdocs.magento.com/guides/v2.4/rest/tutorials/create-custom-rest-api.html) topic.

### CODE REVIEW

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 1]
```diff
@@ -0,0 +1,572 @@
+---
+group: rest-api
+title: Create custom rest api tutorial
+contributor_name: Ziffity
+contributor_link: https://www.Ziffity.com/
+---
+
+## Overview
+
+REST API in Magento 2 defines a set of functions used by the developers to perform requests and get responses using the HTTP protocol. By default, Magento 2 provides some predefined rest APIs like Product API, Order API, Customer data API with which you can take virtual control of everything happening on the site.
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:10]:
```suggestion
The REST API in Magento 2 defines a set of functions used by the developers to perform requests and get responses using the HTTP protocol. By default, Magento 2 provides some predefined rest APIs like Product API, Order API, Customer data API, with which you can take virtual control of everything happening on the site.
```

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+
+Even though Magento 2 has these many rest APIs, they are not enough when it comes to using the custom data and their values. To manage your custom data and fields, you need to create custom rest API in Magento 2. This tutorial describes how you can create custom REST API.
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:12]:
```suggestion
Even though Magento 2 has many REST APIs, they are not enough when it comes to using the custom data and their values. To manage your custom data and fields, you need to create a custom REST API in Magento 2. This tutorial describes how you can create such a custom REST API.
```

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+
+## 1. Create a custom module
+
+Everything starts with a module. `Dev_RestApi` will be used as the namespace:
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:16]:
```suggestion
`Dev_RestApi` will be used as the namespace:
```

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+                    <resource id="Dev_RestApi::products_set_description" title="Set description"
+                              translate="title" sortOrder="20" />
+                </resource>
+            </resource>
+        </resources>
+    </acl>
+</config>
+```
+
+In this tutorial, we have defined acl entries for each endpoints, to have full control over who can access.
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:68]:
```suggestion
In this tutorial, we have defined ACL entries for each endpoint which give full control over who can access the endpoints.
```

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+    <route url="/V1/rest_dev/setDescription" method="PUT">
+        <service class="Dev\RestApi\Api\ProductRepositoryInterface" method="setDescription" />
+        <resources>
+            <resource ref="Dev_RestApi::products_set_description" />
+        </resources>
+    </route>
+</routes>
+```
+
+*  `route.url` is a URL of the endpoint, the full address would be: `<domain>/rest/<store_code><route.url>`
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:93]:
```suggestion
*  `route.url` is the URL of the endpoint, the full address would be: `<domain>/rest/<store_code><route.url>`
```

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+| `PUT` | To update existing object(s) |
+| `DELETE` | To delete object(s) |
+
+*  `service` informs about the interface and the method to be called when the endpoint is reached.
+
+*  `resource` sets the ACL resource that is required to have access to the endpoint. If you want to make it public (no authentication), you can use: `<resource ref="anonymous" />`
+
+## 4. Create Interfaces
+
+In this example, we have created interfaces for request and response
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:110]:
```suggestion
In this example, we have created the interfaces for request and response
```

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+     * @param \Dev\RestApi\Api\RequestItemInterface[] $products
+     * @return void
+     */
+    public function setDescription(array $products);
+}
+```
+
+## 4. Create Models
+
+The objective of models to create classes that implement our interfaces and process data.
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:230]:
```suggestion
The objective of models is to create classes that implement the interfaces and process data.
```

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+In this example, we have created models for request, response and process data. In model we have defined two methods `getItem` which provide product details of the given product id and `setDescription` which updates the description of the given product item(s).
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:231]:
```suggestion
In this example, we have created models for request, response and process data. In each model we have defined two methods `getItem`, which provides product details of the given product id, and `setDescription` which updates the description of the given product item(s).
```

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md [hunk 9]
```diff
@@ -0,0 +1,572 @@
...
+     */
+    public function getId()
+    {
+        return $this->_getData(self::DATA_ID);
+    }
+
+    /**
+     * @return string
+     */
+    public function getDescription()
```
Comment (BarnyShergold) [inline src/guides/v2.3/rest/tutorials/create-custom-rest-api.md:257]: ALL methods should have the return type on the prototype line

**Review by BarnyShergold — CHANGES_REQUESTED** [review]
This looks great! Just a few grammar corrections and a suggestion for the code

Nice new tutorial

File: src/guides/v2.3/rest/tutorials/create-custom-rest-api.md (hunk 8 above)
Comment (nitish-ranjan05) [reply src/guides/v2.3/rest/tutorials/create-custom-rest-api.md]: Hi @BarnyShergold, I have made changes, Kindly review it

**Comment by nitish-ranjan05** [comment]
Hi @BarnyShergold, My apologies. It was clicked on re-request review by mistake.

**Review by BarnyShergold — APPROVED** [review]
Good for me!

**Review by keharper — CHANGES_REQUESTED** [review]
@nitish-ranjan05 Apologies for taking so long to get around to reviewing your PR. This is a good start, but I need you to make the following changes:

* Move the topic to the Extension Best Practices Guide
* Add more linking to topics that discuss each step in detail
* Add more descriptions at the beginning of the steps that explain what you're doing and why
* Please mark all previous review comments as resolved

I'll need to do another review after you've completed this one, but the next one will be much simpler and speedier.

**Comment by dobooth** [comment]
@nitish-ranjan05

**Comment by nitish-ranjan05** [comment]
Hi @dobooth, Apologies. I have been going with a bit busy schedule. I will make it available by next week. Thanks!

**Comment by dobooth** [comment]
Closing this for now. Please reopen it when you are ready to proceed and we will be happy to process it!

**Pull request closed without merging**